### ISSUE TITLE
RISC-V image

### ISSUE THREAD

**Issue by edwardxia** [dart-lang/dart-docker#96]
Now dart has stable support for RISC-V. Having the RISC-V container image would be nice.

**Comment by cpswan** [comment]
It looks like `debian:bullseye-slim` doesn't yet have OS/ARCH support for linux/riscv64 so there's an upstream dependency issue to be dealt with (and although Debian can be run on RISC-V boards/emulators the architecture isn't yet listed as one of their official images).

**Comment by cpswan** [comment]
It also turns out that RISC-V shouldn't have appeared in the beta channel - https://github.com/dart-lang/sdk/issues/50548#issuecomment-1326763431

For my own efforts I think I'm going to hold off until it does make it to beta.

**Comment by cpswan** [comment]
Here's a Dockerfile that I used to make a [build image](https://hub.docker.com/layers/atsigncompany/buildimage/riscv/images/sha256-fc134d576cc75ea4603f6cbfb369a9418974a8d3dafd4c8f5f066045a4d99abc?context=explore) with RISC-V support (based on the existing [beta/bullseye](https://github.com/dart-lang/dart-docker/blob/main/beta/bullseye/Dockerfile) Dockerfile:

```dockerfile
FROM debian:unstable-slim

RUN set -eux; \
    apt-get update; \
    apt-get install -y --no-install-recommends \
        ca-certificates \
        curl \
        dnsutils \
        git \
        openssh-client \
        unzip \
    ; \
    rm -rf /var/lib/apt/lists/*

# Create a minimal runtime environment for executing AOT-compiled Dart code
# with the smallest possible image size.
# usage: COPY --from=dart:xxx /runtime/ /
# uses hard links here to save space
RUN set -eux; \
    case "$(dpkg --print-architecture)" in \
        amd64) \
            TRIPLET="x86_64-linux-gnu" ; \
            FILES="/lib64/ld-linux-x86-64.so.2" ;; \
        armhf) \
            TRIPLET="arm-linux-gnueabihf" ; \
            FILES="/lib/ld-linux-armhf.so.3 \
                /lib/arm-linux-gnueabihf/ld-linux-armhf.so.3";; \
        arm64) \
            TRIPLET="aarch64-linux-gnu" ; \
            FILES="/lib/ld-linux-aarch64.so.1 \
                /lib/aarch64-linux-gnu/ld-linux-aarch64.so.1" ;; \
        riscv64) \
            TRIPLET="riscv64-linux-gnu" ; \
            FILES="/lib/ld-linux-riscv64-lp64d.so.1 \
                /lib/riscv64-linux-gnu/ld-linux-riscv64-lp64d.so.1" ;; \
        *) \
            echo "Unsupported architecture" ; \
            exit 5;; \
    esac; \
    FILES="$FILES \
        /etc/nsswitch.conf \
        /etc/ssl/certs \
        /usr/share/ca-certificates \
        /lib/$TRIPLET/libc.so.6 \
        /lib/$TRIPLET/libdl.so.2 \
        /lib/$TRIPLET/libm.so.6 \
        /lib/$TRIPLET/libnss_dns.so.2 \
        /lib/$TRIPLET/libpthread.so.0 \
        /lib/$TRIPLET/libresolv.so.2 \
        /lib/$TRIPLET/librt.so.1"; \
    for f in $FILES; do \
        dir=$(dirname "$f"); \
        mkdir -p "/runtime$dir"; \
        cp --archive --link --dereference --no-target-directory "$f" "/runtime$f"; \
    done

ENV DART_SDK /usr/lib/dart
ENV PATH $DART_SDK/bin:$PATH

WORKDIR /root
RUN set -eux; \
    case "$(dpkg --print-architecture)" in \
        amd64) \
            DART_SHA256=eaaeee6be87a140a08ae0b6cc76e23ff4e5cb0ef7bbfa8ffa08b90e26b826e6e; \
            SDK_ARCH="x64";; \
        armhf) \
            DART_SHA256=7636eb23c9053bceebf326ae46c8e7be57c5c3cbaedabe2a5a0d6c8717dcc5e9; \
            SDK_ARCH="arm";; \
        arm64) \
            DART_SHA256=b8f3d1f6c65657296757455ac99fab5772dcdb333cc83d15d626717779f2224a; \
            SDK_ARCH="arm64";; \
        riscv64) \
            DART_SHA256=013000bebf117608e9572336f227bffa1839d119753d084009331fc6e1747c9e; \
            SDK_ARCH="riscv64";; \
    esac; \
    SDK="dartsdk-linux-${SDK_ARCH}-release.zip"; \
    BASEURL="https://storage.googleapis.com/dart-archive/channels"; \
    URL="$BASEURL/beta/release/3.0.0-290.3.beta/sdk/$SDK"; \
    echo "SDK: $URL" >> dart_setup.log ; \
    curl -fLO "$URL"; \
    echo "$DART_SHA256 *$SDK" \
        | sha256sum --check --status --strict -; \
    unzip "$SDK" && mv dart-sdk "$DART_SDK" && rm "$SDK" \
        && chmod 755 "$DART_SDK" && chmod 755 "$DART_SDK/bin";
```

**Comment by cpswan** [comment]
The main obstacle to getting this done is Debian support for RISC-V.

As I noted above it's been in 'sid' (unstable) for a little while now, but the Dockerfiles for stable **and** beta start `FROM debian:bullseye-slim`, which is the (now old) stable release (Debian 11).

Debian have announced [riscv64 is now an official architecture](https://lists.debian.org/debian-riscv/2023/07/msg00053.html), but that came after the [release of Debian 12 "Bookworm"](https://www.debian.org/News/2023/20230610) as the latest stable.

If things run their normal course this means that we'll be waiting until Debian 14 "Forky" for RISC-V to make it into a stable release as a supported architecture, and that will be years away (~summer of 2027). I'm hopeful though that we won't have to wait that long, and that we'll start seeing RISC-V show up in Debian 13 "Trixie" and maybe even Debian 12.

As a point of comparison, Ubuntu already have a bunch of 23.04 "Lunar Lobster" releases available for the popular RISC-V dev boards (and QEMU) - [Download Ubuntu for RISC-V Platforms](https://ubuntu.com/download/risc-v), which hints at Ubuntu on RISC-V being ready for the 24.04 LTS release next year. Though there aren't yet any `linux/riscv64` official Ubuntu images on Docker Hub.

**SPECULATION** what I think will happen here is that Ubuntu will fully support RISC-V from 24.04 LTS (including Docker images). Debian will likely then catch up with a 12 stable release (and 13 testing release) that also supports RISC-V. So we have about a year to wait.**/SPECULATION** The reason for optimism is that substantial resource is being poured into RISC-V dependencies from companies (including Google) through the RISC-V Software Ecosystem ([RISE](https://riseproject.dev/)) project. Keep an eye on the [RISE Debian wiki](https://wiki.riseproject.dev/display/HOME/Debian) for updates.

Meanwhile the images here need to catch up to Debian 12...

**Comment by cpswan** [comment]
Returning to this thread (a year and a bit later).

RISC-V Docker images are indeed available in Ubuntu 24.04

![image](https://github.com/user-attachments/assets/740cc542-7e82-4b18-b86a-298642525bba)

and also in Debian 13 "Trixie", which is still at the 'testing' stage and not yet a stable release

![image](https://github.com/user-attachments/assets/0f713c70-9421-447c-9ffa-44d3cf3a07a1)

but no sign of it yet in the stable Debian 12 images, and maybe that will not happen, which leaves us waiting until next year when "Trixie" becomes stable.

**Comment by edwardxia** [comment]
Although still in testing, `debian:trixie` docker image is available now with riscv64 support.

**Comment by cpswan** [comment]
> Although still in testing, `debian:trixie` docker image is available now with riscv64 support.

Have you tried it @edwardxia ? Last time I looked trixie was still missing loads of packages on RISC-V

**Comment by edwardxia** [comment]
Yes, I tried it and it works fine for installing all needed dependencies for this docker image. Note: dnsutils is being renamed to bind9-dnsutils.

**Comment by athomas** [comment]
I just upgraded the base images to trixie. @cpswan is it worth resurrecting https://github.com/dart-lang/dart-docker/pull/140?

**Comment by athomas** [comment]
@cpswan Thanks, the new images should be created with the next release later this week!
